source/game.programme.newsevent.bmx

```diff
@@ -502,33 +504,66 @@ Type TNewsEvent Extends TBroadcastMaterialSource {_exposeToLua="selected"}
 	End Method
 	
 	
-	Method ReplacePlaceholdersFromTemplate:Int(template:TNewsEventTemplate = Null, time:Long = 0)
+	Method ReplacePlaceholdersFromTemplate:Int(template:TNewsEventTemplate = Null, parentVariables:TTemplateVariables = null, time:Long = 0)
 		If not template and not self.templateID Then Return False
 		If not template
 			template = GetNewsEventTemplateCollection().GetByID(self.templateID)
 			If not template Then Return False
 		EndIf
 
+		Local varToUse:TTemplateVariables
+		If templateVariables
+			varToUse = templateVariables
+		Else
+			varToUse = parentVariables
+			If Not varToUse And template.templateVariables
+				'no stored variables and no parent variables - use a copy from the template
+				varToUse = template.templateVariables.Copy()
+			ElseIf template.templateVariables
+				'both template variables and parent variables exist - check if merging is necessary
+				Local mergeNecessary:Int = False
+				'TODO nested variables with parent are not yet considered by this check
+				For Local key:Object = EachIn parentVariables.variables.keys()
+					If Not template.templateVariables.variables.Contains(key)
+						'parent contains additional variable
+						mergeNecessary = True
+						Exit
+					EndIf
+				Next
+				If mergeNecessary
```

Comment:
Wollen wir das auch noch vor dem Merge angehen?

Reply:
Nein. Fachlich sollte ein Mergen von Variablen eigentlich nie nötig sein.
Selbsttriggernde Einzelnachrichten sind behandelt. Bei "normalen" Nachrichtenketten können/sollen alle Variablen in der Hauptnachricht definiert werden.

Der einzie Fall, der mit dem aktuellen Schema nicht behandelt werden kann, ist eine Nachrichtenkette, deren letztes Element wieder die erste Nachricht triggert (Nachrichten"kreis"), da hier dann nie ein Reset der übergebenen Variablen stattfinden würde. Solche Nachrichtenverläufe halte ich aktuell aber auch nicht für nötig.

Reply:
Waere der Kreislauf nicht bspweise DAX-Nachrichten oder "WM-News" (aller 10 Spieltage erneut...) ?

Und waere in dem Fall nicht eher ein "Reset" erwuenscht? Also die Moeglichkeit _alle_ Variablen zurueckzusetzen?

Reply:
Ich habe es gerade nochmal ausprobiert (Mist). Da nun eine Kopie gespeichert wird, funktioniert das Reset auf den Templatevariablen nicht mehr. Mit dem aktuellen Stand bleiben die Börsennachrichten also immer gleich. (Welche WM-Nachrichten meinst Du? Mir war kein anderer Nachrichten-Kreis bekannt.)

Die Idee wäre ja gewesen, dass wann immer eine neue Nachrichtenkette anfängt, diese Initial aus dem Template erzeugt wird, dort die Variablen zurückgesetzt werden.

Ich habe den Commit angepasst. Wenn es gespeicherte Variablen gibt und das Template selbst auch welche hat, aber kein Merge nötig ist (fachlich: wir befinden uns am Anfange einer Nachrichtenkette), werden die Template-Variablen verwendet.

Reply:
... Danke für's Nachhaken.

Reply:
Es gibt noch keine WM-News - war nur als Beispiel gedacht.

> Die Idee wäre ja gewesen, dass wann immer eine neue Nachrichtenkette anfängt, diese Initial aus dem Template erzeugt wird, dort die Variablen zurückgesetzt werden.

Man koennte ja `.Reset()` einfuehren (loescht alle gesetzten Variablen)?
Weiss aber nicht, ob das in der simplen Form von Nutzen waere.

Reply:
Ein Reset gibt es ja schon. Die Frage wäre, woran man erkennen soll, dass es aufgerufen werden muss.
Wenn die Nachrichtenagentur komplett neue Nachrichten aus der TemplateCollection holt, macht sie schon genau so ein Reset.
Wenn eine Nachrichtenkette sich aber selbst triggert, gibt es keinen solchen Startpunkt mehr. Dafür müsste man die Syntax erweitern (triggerNewsEventWithVariableRest). Das wollte ich aber vermeiden.

Mit der Konvention "Allein die Startnachricht darf Variablen enthalten" hätte man die Information und die aktuelle Implementierung würde funktionieren (gespeicherte und Templatevariablen, aber kein Merge nötig).

Reply:
> "Allein die Startnachricht darf Variablen enthalten"

Eine News kann auch eine andere (fremde) Newskette anstossen. Da die eine neue Startnews ist, sollte das gehen.
Allerdings kann in einer bestehenden Newskette keine Variable _optional_ "ueberschreiben" 

News1 - Streik in %CITY%
| -> News 2 (Optional!) - Streik zieht nach %CITY% um. (Variable ueberschreiben)
News3 -> Streiks in %CITY% weiten sich aus.


Ich hatte auch ueberlegt, ob eine Nachricht einen Flag bekommen kann, der Variablen "zuruecksetzt" (Reset() ausloest).

Reply:
Den Fall andere Nachrichtenketten anstoßen (in Verbindung mit Variablen), muss man sich im konkreten Fall anschauen. Ich denke, es funktioniert im Moment noch nicht (ohne Erweiterung von Syntax und trigger-Logik).

Für erste (weitreichende) Verbesserungen der Nachrichten sollte die aktuelle Logik schon taugen: (einfache) Nachrichtenketten mit Ende, die von der "Nächste-Nachrichten"-Logik erneut gestartet werden können.